ui/lesson-renderer/src/main/java/org/cru/godtools/tool/lesson/ui/LessonPageAdapter.kt

```diff
@@ -43,21 +47,44 @@ class LessonPageAdapter @AssistedInject internal constructor(
             notifyDataSetChanged()
         }
 
+    private var primaryItemObserver: PrimaryItemChangeObserver<*>? = null
+
     override fun getItemCount() = pages.size
     private fun getItem(position: Int) = pages[position]
     override fun getItemId(position: Int) = IdUtils.convertId(getItem(position).id)
 
     // region Lifecycle
-    override fun onCreateViewDataBinding(parent: ViewGroup, viewType: Int) =
-        LessonPageBinding.inflate(LayoutInflater.from(parent.context), parent, false).apply {
-            callbacks = _callbacks
-            bindController(controllerFactory)
+    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
+        primaryItemObserver = onUpdatePrimaryItem(recyclerView) { primary, previous ->
+            previous?.binding?.controller?.lifecycleOwner?.apply { maxState = minOf(maxState, Lifecycle.State.STARTED) }
+            primary?.binding?.controller?.lifecycleOwner?.apply { maxState = maxOf(maxState, Lifecycle.State.RESUMED) }
```

Comment:
Could you explain what the ```maxState``` is? Also to confirm what I "believe" this is doing, the ```primary``` and ```previous``` refers to the current and last page, and you are binding a specific state to the individual lifecycles of those pages?

Reply:
relevant class: https://github.com/CruGlobal/android-gto-support/blob/master/gto-support-androidx-lifecycle/src/main/java/org/ccci/gto/android/common/androidx/lifecycle/ConstrainedStateLifecycleOwner.kt

Basically the `ConstrainedStateLifecycleOwner` has a `maxState` property the limits the maximum state the lifecycle can reach. Whenever the "primaryItem" is updated, the previous primary item is constrained to a max state of `STARTED`, and the new primary item is constrained to a max state of `RESUMED`.

I use the `minOf` and `maxOf` functions to not potentially pave over a more restrictive `maxState` set elsewhere, although I don't believe there are any code paths that would make that necessary